h/schemas/forms/admin/group.py

```diff
@@ -24,56 +22,65 @@
+    group and the group's organization.
+    """
+    user_svc = form.bindings["user_svc"]
 
-        """
-        user_svc = kw["user_svc"]
+    # Get the authority from the list of organisations based on the one
+    # actually picked in the form
+    authority = form.bindings["organizations"][value["organization"]].authority
 
-        # A {pubid: models.Organization} dict of all the organizations
-        # available to choose from in the form.
-        organizations = kw["organizations"]
+    exc = colander.Invalid(form, None)
 
-        # The pubid of the organization that the user has selected in the form.
-        selected_pubid = value["organization"]
+    _validate_members(form, value["members"], authority, user_svc, exc)
+    _validate_creator(value["creator"], authority, user_svc, exc)
 
-        # The models.Organization object for the selected organization.
-        selected_organization = organizations[selected_pubid]
+    if exc.children:
+        raise exc
 
-        # The authority that the new group will belong to if it is created.
-        authority = selected_organization.authority
 
-        # The username string that was entered for the group creator.
-        creator_username = value["creator"]
+def _validate_members(form, members_value, authority, user_svc, parent_exc):
+    if not members_value:
+        return
 
-        # The models.User object for the group creator user, or None.
-        user = user_svc.fetch(creator_username, authority)
+    members_node = form.get("members")
+    member_node = members_node.get("member")
+    members_error = colander.Invalid(members_node, None)
 
-        if not user:
-            # Either the username doesn't exist at all, or it has a different
-            # authority than the chosen organization.
-            exc = colander.Invalid(form, _("User not found"))
-            exc["creator"] = _(
-                "User {creator} not found at authority {authority}"
-            ).format(creator=creator_username, authority=authority)
-            raise exc
+    for pos, node_value in enumerate(members_value):
+        if not user_svc.fetch(node_value, authority):
+            # If we don't attach errors in the right position, they don't
+            # appear. For a list type value the correct position is the index
+            # of the value in the list
+            members_error.add(
+                colander.Invalid(
+                    member_node,
+                    _("User '{username}' not found at authority '{authority}'").format(
+                        username=node_value, authority=authority
+                    ),
+                ),
+                pos=pos,
+            )
 
-    return validate
+    if members_error.children:
+        # For non-lists the correct position is the index within the flat list
+        # of children that colander stores internally for the parent node
+        parent_exc.add(members_error, pos=form.children.index(members_node))
```

Comment:
By this point, it's pretty easy to lose track of what `non-list` is in context here: `parent_exc` is, at least in _practice_ here, any error(s) associated with the form itself? There's about one level of abstraction too many for my little beady brain's comfort here, but it's not catastrophic.

Reply:
No snappy way I can think of to explain this...

From what I understand the structure of the errors you produce must _mostly_ reflect the structure of the schema. The issue is that the schema has a single node representing the structure of all entries in a list. So the schema be like:

```
root
    creator (pos:0)
    members (pos:1)
        member (pos:0)  # represents many, but is a single thing
```

Where as the data is like:

```
root
    creator (pos:0)
    members (pos:1)
          member.0 (pos:0)
          member.1 (pos:1)
          member.2 (pos:2)
```

To add an error about the creator we need a root error, which we then add an error to in the 0th position, reflecting it's place in the schema:

```
root = Invalid(form_node, None)
creator = Invalid(creator_node, "blah") 

root.add(creator, 0)  # 0 is the position in the schema
```

To add an error about a specific member it's different. The first part is the same, we need to nest a members parent error to reflect the position of the members node in the schema, but we then switch to using the value position in the data for the many values for member:

```
root = Invalid(form_node, None)
members = Invalid(members_node, None)
member = Invalid(member_node, "Wrong username")

root.add(members, 1)  # 1 is the position in the schema
members.add(member, 2) # 2 here is the position in the values
```

----

Long story short, both are individual errors about a part of the form, but one is an error about a single valued field: the other is an error about one value within a multi valued list field.